tfx_addons/sampling/executor.py

```diff
@@ -0,0 +1,286 @@
+          splits as specified in splits. May also include copies of the
+          other non-sampled spits, as specified by null_classes.
+      exec_properties: A dict of execution properties, including:
+        - name: Optional unique name. Necessary if multiple components are
+          declared in the same pipeline.
+        - label: The name of the column containing class names to sample by.
+        - splits: A list containing splits to sample.
+        - copy_others: Determines whether we copy over the splits that aren't
+          sampled, or just exclude them from the output artifact. Defaults
+          to True.
+        - shards: The number of files that each sampled split should
+          contain. Default 0 is Beam's tfrecordio function's default.
+        - null_classes: A list determining which classes that we should
+          not sample. Defaults to None.
+        - sampling_strategy: An enum of type SamplingStrategy, determining if
+          the executor should over or undersample.
+    Returns:
+      None
+    """
+
+    self._log_startup(input_dict, output_dict, exec_properties)
+
+    label = exec_properties[spec.SAMPLER_LABEL_KEY]
+    sampling_strategy = exec_properties[spec.SAMPLER_SAMPLE_KEY]
+    splits = json_utils.loads(exec_properties[spec.SAMPLER_SPLIT_KEY])
+    copy_others = exec_properties[spec.SAMPLER_COPY_KEY]
+    shards = exec_properties[spec.SAMPLER_SHARDS_KEY]
+    null_classes = json_utils.loads(exec_properties[spec.SAMPLER_CLASSES_KEY])
+
+    input_artifact = artifact_utils.get_single_instance(
+        input_dict[spec.SAMPLER_INPUT_KEY])
+    output_artifact = artifact_utils.get_single_instance(
+        output_dict[spec.SAMPLER_OUTPUT_KEY])
+
+    if sampling_strategy not in (spec.SamplingStrategy.UNDERSAMPLE,
+                                 spec.SamplingStrategy.OVERSAMPLE):
+      raise ValueError("Invalid sampling strategy!")
+
+    if not splits:
+      raise ValueError("No splits are marked for sampling!")
+
+    for split in splits:
+      if split not in input_artifact.split_names:
+        raise ValueError(
+            f"Invalid split name {split} is not in input artifact!")
+
+    if shards < 0:
+      raise ValueError("Shards value must be non-negative!")
+
+    if copy_others:
+      output_artifact.split_names = input_artifact.split_names
+    else:
+      output_artifact.split_names = artifact_utils.encode_split_names(splits)
+
+    # Fetch the input uri for each split
+    split_data = {}
+    for split in artifact_utils.decode_split_names(input_artifact.split_names):
+      uri = artifact_utils.get_split_uri([input_artifact], split)
+      split_data[split] = uri
+
+    for split, uri in split_data.items():
+      if split in splits:  # Undersampling split
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        split_dir = os.path.join(output_dir, f"Split-{split}")
+        with self._CreatePipeline(split_dir) as p:
```

Comment:
@RossKohler From the executor, it should be inheriting the beam pipeline args. Do you have some repro code?